### ISSUE TITLE
HTTPError: 400 Client Error: Bad Request for url: <filter messages toRecipients>

### ISSUE THREAD

**Issue by olivierreberga** [O365/python-o365#452]
I'm trying to filter emails by "to" with the following:

```
query = mailbox.new_query() 
query = query.on_attribute('subject').equals('BlaBla')
query = query.chain('and').on_attribute('from').equals('aaaa@gmail.com')
query = query.chain('and').on_attribute('to').equals('bbbb@gmail.com')
```

But this raises an error:

`requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://graph.microsoft.com/v1.0/me/messages?%24top=25&%24filter=subject+eq+%27Second+message%27+and+from%2FemailAddress%2Faddress+eq+%27aaaa%40gmail.com%27+and+toRecipients%2FemailAddress%2Faddress+eq+%27bbbb%40gmail.com%27 | Error Message: Invalid filter clause`

It's working fine if i don't use the "to" argument

**Comment by janscas** [comment]
The query generated by the Query class seems good to me.
MS Graph has a lot of limitations on what queries can be performed.
And what's worst is that these limitations aren't described anywhere...

So you have to try. Try to filter just using the TO filter and see if it works, then add another one, etc...
If they work independently but not together that's because is a filter not allowed by MS Graph.

Let me know if the independent filters don't work.

**Comment by olivierreberga** [comment]
I think there is an issue with the filter on "to" criteria, the following request:

```
query = mailbox.new_query()  # you can use the shorthand: mailbox.q()
query = query.on_attribute('to').equals('bbbb@gmail.com')
```

raises the same error:

```
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://graph.microsoft.com/v1.0/me/messages?%24top=25&%24filter=toRecipients%2FemailAddress%2Faddress+eq+%27bbbb%40gmail.com%27 | Error Message: Invalid filter clause
```
Is that the right way to filter on "to"?
Because "to" is a list of recipients whereas from is only one value.

**Comment by janscas** [comment]
As I told you, the docs are very unreliable.

To is not filterable:
https://stackoverflow.com/questions/39753969/unable-to-filter-messages-by-recipient-in-microsoft-graph-api-one-or-more-inval

**Comment by olivierreberga** [comment]
Thanks, as a workaround, the search function cal be used to filter on to

**Comment by janscas** [comment]
Yes, user `Query.search`. but note that you will not be able to filter by the other clauses.

**Comment by chrisspen** [comment]
Doesn't `search()` support filtering by specific fields using an `key:value` syntax? You can filter on the `to` field by doing, `query.search("to:sldkfjsld@lsdkjlfds.com")`

**Comment by janscas** [comment]
https://docs.microsoft.com/en-us/graph/query-parameters?context=graph%2Fapi%2F1.0&view=graph-rest-1.0#search-parameter

**Comment by chrisspen** [comment]
The other big downside of search() is it doesn't support all the normal pagination features. Notably it looks like you can't use `$skip` so you can't query specific slices. You can break up the query into batches, but you can't request, say, batch 10, without also downloading batches 0-9 first.
